utils/writeable/src/try_writeable.rs

```diff
@@ -0,0 +1,442 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use core::cmp::Ordering;
+
+/// A writeable object that can fail while writing.
+///
+/// The default [`Writeable`] trait returns a [`fmt::Error`], which originates from the sink.
+/// In contrast, this trait allows the _writeable itself_ to trigger an error.
+///
+/// Implementations are expected to always make a _best attempt_ at writing to the sink
+/// and should write replacement values in the error state. Therefore, [`TryWriteable::Error`]
+/// can be safely ignored to emulate a "lossy" mode.
+///
+/// Any error substrings should be annotated with [`Part::ERROR`].
+///
+/// # Implementer Notes
+///
+/// This trait requires that implementers make a _best attempt_ at writing to the sink,
+/// _even in the error state_, such as with a placeholder or fallback string.
+///
+/// In [`TryWriteable::try_write_to_parts()`], error substrings should be annotated with
+/// [`Part::ERROR`]. Becuause of this, writing to parts is not default-implemented like
+/// it is on [`Writeable`].
+///
+/// Furthermore, [`TryWriteable::try_writeable_length_hint()`] is not default-implemented because
+/// it has an invariant that the error state should be correctly propagated.
+///
+/// The trait is implemented on [`Result<T, E>`] where `T` and `E` both implement [`Writeable`];
+/// In the `Ok` case, `T` is written, and in the `Err` case, `E` is written as a fallback value.
+/// This impl, which writes [`Part::ERROR`], can be used as a basis for more advanced impls.
+///
+/// # Examples
+///
+/// Implementing on a custom type:
+///
+/// ```
+/// use core::fmt;
+/// use writeable::LengthHint;
+/// use writeable::PartsWrite;
+/// use writeable::TryWriteable;
+///
+/// #[derive(Debug, PartialEq, Eq)]
+/// enum MyWriteableError {
+///     MissingName,
+/// }
+///
+/// #[derive(Debug, PartialEq, Eq)]
+/// struct HelloWorldWriteable {
+///     pub name: Option<&'static str>,
+/// }
+///
+/// impl TryWriteable for HelloWorldWriteable {
+///     type Error = MyWriteableError;
+///
+///     fn try_write_to_parts<S: PartsWrite + ?Sized>(
+///         &self,
+///         sink: &mut S,
+///     ) -> Result<Result<(), Self::Error>, fmt::Error> {
+///         sink.write_str("Hello, ")?;
+///         // Use `impl TryWriteable for Result` to generate the error part:
+///         let _ = self.name.ok_or("nobody").try_write_to_parts(sink)?;
+///         sink.write_char('!')?;
+///         // Return a doubly-wrapped Result.
+///         // The outer Result is for fmt::Error, handled by the `?`s above.
+///         // The inner Result is for our own Self::Error.
+///         if self.name.is_some() {
+///             Ok(Ok(()))
+///         } else {
+///             Ok(Err(MyWriteableError::MissingName))
+///         }
```

Comment:
Let's not repeat the same check

```suggestion
///         let result = self.name.ok_or("nobody").try_write_to_parts(sink)?;
///         sink.write_char('!')?;
///         Ok(result.map_err(|_| MyWriteableError::MissingName))
```

Reply:
Sure, this makes the code shorter, but is it more clear? I wrote it the more lengthy way in an attempt to make it more clear, especially regarding the doubly nested results. I use the concise syntax in most other places, just not in the docs test / impl example.

Reply:
This is not about conciseness. I think it's weird that you do all the writing, discard the error you produce on the way, and then after you're done you compute the error from scratch again.

Reply:
I changed it to not discard the error.